Fix expecttest accept on Python 3.8 and later

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59760 Dispatch to Python via __torch_dispatch__
* #59758 Make PyObject_FastGetAttrString accept const char*
* **#59709 Fix expecttest accept on Python 3.8 and later**

Fixes #59705.

Python 3.8 fixed tracebacks to report the beginning of the line
that raised an error, rather than the end.  This makes for a simpler
implementation (no more string reversing) but need to actually
implement.  This wasn't caught by tests because we hard coded line
numbers to do substitutions, so I also added a little smoketest to
detect future changes to traceback line number behavior.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28994919](https://our.internmc.facebook.com/intern/diff/D28994919)

ezyang committed:
Fix expecttest accept on Python 3.8 and later

Fixes #59705.

Python 3.8 fixed tracebacks to report the beginning of the line
that raised an error, rather than the end.  This makes for a simpler
implementation (no more string reversing) but need to actually
implement.  This wasn't caught by tests because we hard coded line
numbers to do substitutions, so I also added a little smoketest to
detect future changes to traceback line number behavior.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28994919).

ezyang commented:
Note to reviewers, I don't think either of you are experts in this code, I hope you can just review it on its merits.

bdhirsh commented on torch/testing/_internal/expecttest.py:
```diff
@@ -152,54 +168,85 @@ def replace_string_literal(src : str, lineno : int,
+                            ])
+
+        return (src[:i] + RE_EXPECT.sub(replace, src[i:], count=1), delta[0])
+    else:
```
more of these two conditions could probably be de-duped? Not a big deal though

walterddr commented on test/test_expecttest.py:
```diff
@@ -96,6 +99,72 @@ def test_sample(self):
 ''')
 """)
 
+    def test_sample_lineno_at_start(self):
```
nit: 
maybe specifically named previous test in line:49 as 
`test_sample_lineno_at_end(self)`

ezyang committed:
Update on "Fix expecttest accept on Python 3.8 and later"


Fixes #59705.

Python 3.8 fixed tracebacks to report the beginning of the line
that raised an error, rather than the end.  This makes for a simpler
implementation (no more string reversing) but need to actually
implement.  This wasn't caught by tests because we hard coded line
numbers to do substitutions, so I also added a little smoketest to
detect future changes to traceback line number behavior.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28994919](https://our.internmc.facebook.com/intern/diff/D28994919)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28994919).
